### DISCUSSION TITLE
Object returned from useMap() missing addSource() and addLayer() methods in maplibre-based map

### DISCUSSION THREAD

**Discussion by wdawson61** [uber/react-map-gl#2233] [Q&A] [answered]
I'm trying to create a <BingMap/> component using a react-map-gl <Map> with a custom style pointing to raster satellite tiles from Bing maps.   That's working, but a child component of the map,  tries to use addImage() addSource() and addLayer() on the value returned from useMap().  The addSource() method seems to be missing from the returned object.  Is this correct?  This is a Maplibre map.

**Comment by Pessimistress** [comment, marked as answer]
From the [documentation](https://visgl.github.io/react-map-gl/docs/api-reference/map#methods):

> The [MapRef](https://visgl.github.io/react-map-gl/docs/api-reference/types#mapref) object exposes [Map methods](https://docs.mapbox.com/mapbox-gl-js/api/map/#map-instance-members) that are safe to call without breaking the React bindings. For example, `setStyle()` is hidden from the ref object, because the style is supposed to be changed by updating the `mapStyle` prop. Calling the method directly may cause the the React prop to mismatch with the underlying state, and lead to unexpected behaviors.

> You can still access the hidden members via `getMap()`

**Reply by wdawson61** [reply]
Thanks.  I've moved to using the reactive implementations:
`<Source > <Layer /></Source>`
and all is working on that front.  Thanks for the help!
